In [82]:
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_california_housing
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,GradientBoostingRegressor  
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score

In [26]:
house=fetch_california_housing()
house_df=pd.DataFrame(house.data,columns=house.feature_names)
house_df['Target']=house.target
house_df.head(10)

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
5,4.0368,52.0,4.761658,1.103627,413.0,2.139896,37.85,-122.25,2.697
6,3.6591,52.0,4.931907,0.951362,1094.0,2.128405,37.84,-122.25,2.992
7,3.1200,52.0,4.797527,1.061824,1157.0,1.788253,37.84,-122.25,2.414
8,2.0804,42.0,4.294118,1.117647,1206.0,2.026891,37.84,-122.26,2.267
9,3.6912,52.0,4.970588,0.990196,1551.0,2.172269,37.84,-122.25,2.611


In [27]:
house_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
 8   Target      20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


In [28]:
house_df.isnull().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
Target        0
dtype: int64

In [29]:
scaler=StandardScaler()

Scaled_house=scaler.fit_transform(house_df)
scaled_df=pd.DataFrame(Scaled_house,columns=house_df.columns)

scaled_df.head(10)

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,2.344766,0.982143,0.628559,-0.153758,-0.974429,-0.049597,1.052548,-1.327835,2.129631
1,2.332238,-0.607019,0.327041,-0.263336,0.861439,-0.092512,1.043185,-1.322844,1.314156
2,1.782699,1.856182,1.155620,-0.049016,-0.820777,-0.025843,1.038503,-1.332827,1.258693
3,0.932968,1.856182,0.156966,-0.049833,-0.766028,-0.050329,1.038503,-1.337818,1.165100
4,-0.012881,1.856182,0.344711,-0.032906,-0.759847,-0.085616,1.038503,-1.337818,1.172900
5,0.087447,1.856182,-0.269730,0.014669,-0.894071,-0.089618,1.038503,-1.337818,0.544611
6,-0.111366,1.856182,-0.200918,-0.306633,-0.292712,-0.090725,1.033821,-1.337818,0.800259
7,-0.395137,1.856182,-0.255232,-0.073542,-0.237079,-0.123476,1.033821,-1.337818,0.299362
8,-0.942359,1.061601,-0.458703,0.044254,-0.193810,-0.100499,1.033821,-1.342809,0.171971
9,-0.094470,1.856182,-0.185283,-0.224687,0.110844,-0.086501,1.033821,-1.337818,0.470083


<h3>Created a DataFrame:</h3><p>Converted the California Housing data into a table format (Pandas DataFrame) to make it much easier to read and work with.</p>

<h3>Checked for Missing Data:</h3><p>Looked for any blank and confirmed the dataset was already 100% complete with no missing values. Finding these early prevents errors and keeps the machine learning model accurate.</p>

<h3>Scaled the Data</h3><p>Used StandardScaler to put all the numbers onto a similar scale, making sure no single feature accidentally overpowers the others</p>

In [34]:
x=scaled_df.drop("Target",axis=1)
y=scaled_df["Target"]

In [36]:
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=42,test_size=0.2)

# Linear Regression Model

In [38]:
lr=LinearRegression()
lr.fit(x_train,y_train)

LinearRegression()

In [49]:
y_pred_lr=lr.predict(x_test)

In [84]:
print("mse=",mean_squared_error(y_test,y_pred_lr))
print("mae=",mean_absolute_error(y_test,y_pred_lr))
print("R2_score=",r2_score(y_test,y_pred_lr))

mse= 0.4174769805249499
mae= 0.46207392160821764
R2_score= 0.575787706032451


# Decision Tree Regressor  

In [58]:
dt=DecisionTreeRegressor(random_state=42)
dt.fit(x_train,y_train)

DecisionTreeRegressor(random_state=42)

In [ ]:
y_pred_dt=dt.predict(x_test)

In [85]:
print("mse=",mean_squared_error(y_test,y_pred_dt))
print("mae=",mean_absolute_error(y_test,y_pred_dt))
print("R2_score=",r2_score(y_test,y_pred_dt))

mse= 0.3687217921753256
mae= 0.39284221172623507
R2_score= 0.6253294802078968


# Random Forest Regressor  

In [67]:
rf=RandomForestRegressor(random_state=42,n_estimators=100)
rf.fit(x_train,y_train)

RandomForestRegressor(random_state=42)

In [68]:
y_pred_rf=rf.predict(x_test)

In [86]:
print("mse=",mean_squared_error(y_test,y_pred_rf))
print("mae=",mean_absolute_error(y_test,y_pred_rf))
print("R2_score=",r2_score(y_test,y_pred_rf))

mse= 0.19265073556179227
mae= 0.2838078315150398
R2_score= 0.8042411575257635


# Gradient Boosting Regressor  

In [72]:
gb=GradientBoostingRegressor(random_state=42)
gb.fit(x_train,y_train)

GradientBoostingRegressor(random_state=42)

In [73]:
y_pred_gb=gb.predict(x_test)

In [88]:
print("mse=",mean_squared_error(y_test,y_pred_gb))
print("mae=",mean_absolute_error(y_test,y_pred_gb))
print("R2_score=",r2_score(y_test,y_pred_gb))

mse= 0.22079452229981855
mae= 0.3220741526056764
R2_score= 0.7756433164710084


# Support Vector Regressor (SVR)  

In [77]:
svr=SVR()
svr.fit(x_train,y_train)

SVR()

In [78]:
y_pred_svr=svr.predict(x_test)

In [89]:
print("mse=",mean_squared_error(y_test,y_pred_svr))
print("mae=",mean_absolute_error(y_test,y_pred_svr))
print("R2_score=",r2_score(y_test,y_pred_svr))

mse= 0.2643983636325156
mae= 0.34331371235964975
R2_score= 0.7313359979350706


<h1>Conclusion</h1><p>After implementing all the regression models, the Random Forest Regressor gave the best performance with the highest R² score of 0.80. It predicted house prices more accurately because it can handle complex and non-linear data patterns effectively.

The Linear Regression model gave the lowest R² score. This is because it works mainly for linear relationships and could not capture the complex patterns in the California Housing dataset.

Therefore, Random Forest Regressor was the best model for this dataset, while Linear Regression was the least effective model.</p>

<h2><p>RandomForest>GradientBoosting>SupportVector>DcisionTree>LinearRegression</p></h2>